### PR TITLE
Introduce GradientCheckpointingLayer

### PR DESCRIPTION
**Pull request by qubvel**
# What does this PR do?

A super minimal abstraction for a layer with gradient checkpointing that keeps the logic for enabling and disabling gradient checkpointing within PreTrainedModel for backward compatibility. It allows for a gradual rollout of the feature by supporting both checkpointing mechanisms: with a the current wrap of `_gradient_checkpointing_func` and using inheritance from `GradientCheckpointingLayer`.

I've applied this to Llama, but it's just a PoC for the discussion. Perhaps it's better to start with another less popular model that has fewer dependent models to see how it goes and check if it can be breaking for the hub custom code

### CODE REVIEW

**Comment by qubvel** [comment]
run-slow: llama

**Review by ArthurZucker — APPROVED** [review]
Nice! Lets put it in another file to separate a bit! Otherwise marvelleous

**Comment by qubvel** [comment]
Ended up applying to all llama-based models and SigLIP/2 for the first iteration, all relevant tests pass

```
RUN_SLOW=1 pytest -k "gradient" tests/models/
```

CI error is unrelated

cc @ArthurZucker to merge if OK for you

**Review by ArthurZucker — APPROVED** [review]
🧼 thanks a lot!
